File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RingBuffer.java
package com.mapbox.services.android.navigation.v5.utils;

import androidx.annotation.IntRange;

import java.util.ArrayDeque;
import java.util.Collection;

public class RingBuffer<T> extends ArrayDeque<T> {

  private final int maxSize;

  public RingBuffer(@IntRange(from = 0) int maxSize) {
    this.maxSize = maxSize;
  }

  @Override
  public boolean add(T item) {
    boolean result = super.add(item);
    resize();
    return result;
  }

  @Override
  public void addFirst(T item) {
    super.addFirst(item);
    resize();
  }

  @Override
  public void addLast(T item) {
    super.addLast(item);
    resize();
  }

  @Override
  public boolean addAll(Collection<? extends T> collection) {
    boolean result = super.addAll(collection);
    resize();
    return result;
  }

  @Override
  public void push(T item) {
    super.push(item);
    resize();
  }

  private void resize() {
    while (size() > maxSize) {
      pop();
    }
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/utils/RingBuffer.kt
package com.mapbox.services.android.navigation.v5.internal.utils

import androidx.annotation.IntRange
import java.util.ArrayDeque

class RingBuffer<T>(
    @IntRange(from = 0) private val maxSize: Int
) : ArrayDeque<T>() {

    override fun add(element: T): Boolean {
        val result = super.add(element)
        resize()
        return result
    }

    override fun addFirst(element: T) {
        super.addFirst(element)
        resize()
    }

    override fun addLast(element: T) {
        super.addLast(element)
        resize()
    }

    override fun addAll(elements: Collection<T>): Boolean {
        val result = super.addAll(elements)
        resize()
        return result
    }

    override fun push(element: T) {
        super.push(element)
        resize()
    }

    private fun resize() {
        while (size > maxSize) {
            pop()
        }
    }
}
